pennylane_rigetti/qc.py

```diff
@@ -192,13 +188,14 @@ def apply(self, operations, **kwargs):
 
         rotations = kwargs.get("rotations", [])
         self.prog += self.apply_rotations(rotations)
-
-        qubits = sorted(self.wiring.values())
-        ro = self.prog.declare("ro", "BIT", len(qubits))
-        for i, q in enumerate(qubits):
-            self.prog.inst(MEASURE(q, ro[i]))
-
         self.prog.wrap_in_numshots_loop(self.shots)
+        # Measure every qubit used by the program into a readout register.
+        # Devices don't always have sequentially adressed qubits, so
+        # we use a normalized value to index them into the readout register.
+        used_qubits = self.prog.get_qubits(indices=True)
+        ro = self.prog.declare("ro", "BIT", len(used_qubits))
+        for i, qubit in enumerate(used_qubits):
+            self.prog += MEASURE(qubit, ro[i])
```

Comment:
This is a large departure from how the plugin was originally written so I want to draw extra attention to this and solicit some extra feedback and clarification.

Is there any reason we should still be measuring every qubit here?

The `ro` register we declare here is ultimately what gets returned by `generate_samples`. What is the expected shape of the matrix `generated_samples` returns? Should it relate to the number of qubits a circuit actually uses, or maybe the number of qubits on a circuit that are measured?

Reply:
In PennyLane, the shape of the output of a circuit depends on the number of qubits **measured**, and not the number of qubits **used** in a circuit. Here is a dummy example:
```python
import pennylane as qml

dev = qml.device("default.qubit", wires=4, shots=1000)


@qml.qnode(dev)
def circuit():
    qml.PauliX(0)
    qml.PauliX(1)
    qml.PauliX(2)
    return qml.probs(op=qml.PauliZ(0) @ qml.PauliZ(1))


print(qml.math.shape(circuit()))
```
Which returns: `(4,)` instead of `(8,)`.

Reply:
`generate_samples` uses all the qubits of the **device**! Thus if a device has 10 qubits, and the circuit executed uses 5 qubits and measures 3, we still return 10 qubits in `generate_samples` (I know, it is not very optimal).

Reply:
Gotcha, I'll fit the data into a register of that size then. Based on the old code, it looks like it's expected that each row of samples should be ordered from the lowest valued qubit label to the highest valued qubit label, so I went with that pattern. Let me know if it should be something different though!